consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/statemachine/QbftBlockHeightManager.java

```diff
@@ -130,19 +131,48 @@ public void handleBlockTimerExpiry(final ConsensusRoundIdentifier roundIdentifie
 
     logValidatorChanges(qbftRound);
 
+    if (roundIdentifier.equals(qbftRound.getRoundIdentifier())) {
+      buildBlockAndMaybePropose(roundIdentifier, qbftRound);
+    } else {
+      LOG.trace(
+              "Block timer expired for a round ({}) other than current ({})",
+              roundIdentifier,
+              qbftRound.getRoundIdentifier());
+    }
+  }
+
+  private void buildBlockAndMaybePropose(
+          final ConsensusRoundIdentifier roundIdentifier, final QbftRound qbftRound) {
+
     // mining will be checked against round 0 as the current round is initialised to 0 above
     final boolean isProposer =
-        finalState.isLocalNodeProposerForRound(qbftRound.getRoundIdentifier());
-
-    if (isProposer) {
-      if (roundIdentifier.equals(qbftRound.getRoundIdentifier())) {
-        final long headerTimeStampSeconds = Math.round(clock.millis() / 1000D);
-        qbftRound.createAndSendProposalMessage(headerTimeStampSeconds);
+            finalState.isLocalNodeProposerForRound(qbftRound.getRoundIdentifier());
+
+    final long headerTimeStampSeconds = Math.round(clock.millis() / 1000D);
+    final Block block = qbftRound.createBlock(headerTimeStampSeconds);
+    final boolean blockHasTransactions = !block.getBody().getTransactions().isEmpty();
+    if (blockHasTransactions) {
+      if (isProposer) {
+        LOG.debug("Block Has Transactions and I AM proposer so send proposal");
+        qbftRound.sendProposalMessage(block);
+      }else{
+        LOG.debug("Block Has Transactions and I am NOT proposer");
+      }
+    } else {
+      final long emptyBlockPeriodSeconds = finalState.getBlockTimer().getEmptyBlockPeriodSeconds();
+      final long emptyBlockPeriodExpiryTime = parentHeader.getTimestamp() + emptyBlockPeriodSeconds;
+      final long nowInSeconds = finalState.getClock().millis() / 1000;
+      if (nowInSeconds >= emptyBlockPeriodExpiryTime) {
+        if (isProposer) {
+          LOG.debug("Block Do Not Have Transactions and I AM proposer so send proposal");
+          qbftRound.sendProposalMessage(block);
+        } else {
+          LOG.debug("Block Do Not Have Transactions and I am NOT proposer");
+        }
       } else {
-        LOG.trace(
-            "Block timer expired for a round ({}) other than current ({})",
-            roundIdentifier,
-            qbftRound.getRoundIdentifier());
+        finalState.getRoundTimer().cancelTimer();
```

Comment:
Can you explain the purpose of cancelling the round timer here?

Reply:
I need to get back to this one and check my changes history.....

Reply:
Well, it's quite some time I did that code, don't really remember what made me add it, but if i remove it I start getting this error around the same time a block is created...

> 2024-04-16 19:50:35.026+01:00 | BftProcessorExecutor-QBFT-0 | ERROR | QbftBlockHeightManager | Received Round timer expiry before round is created timerRound=ConsensusRoundIdentifier{Sequence=11, Round=0}

Reply:
Hi @siladu , checking again I did find out that cancelling come from... your code 🥇 

```
        finalState.getRoundTimer().cancelTimer();
        long blockPeriodInMillis = 5_000L;
        final long newExpiry = nowInMillis + blockPeriodInMillis;
        finalState.getBlockTimer().startTimer(roundIdentifier, newExpiry);
        currentRound = Optional.empty();
```
so since it was working fine 💯 I never look into it...

Reply:
@amsmota haha I don't remember why did that :)

However, my commit message gives some clue... [`3c388d3` (#4634)](https://github.com/hyperledger/besu/pull/4634/commits/3c388d3cc023268013eb6a98688b5e5d297c855b)
> Delay start of round by cancelling roundChangeTimer and adding short delay to a new BlockTimer to create a loop
In that loop (In QbftBlockHeightManager), check if emptyBlockPeriod has been surpassed.